src/main/java/buildcraft/factory/BlockTank.java

```diff
@@ -88,123 +88,210 @@ public IIcon getIconAbsolute(IBlockAccess iblockaccess, int i, int j, int k, int
-
-                                return true;
-                            }
-                        }
-                    }
-                } else if (current.getItem() instanceof IFluidContainerItem) {
-                    if (current.stackSize != 1) {
-                        return false;
+                        player.inventory.setInventorySlotContents(player.inventory.currentItem, emptied);
                     }
+                }
+                syncPlayerAndTanks(world, x, y, z, player, clicked);
+                return true;
+            }
 
-                    if (!world.isRemote) {
-                        IFluidContainerItem container = (IFluidContainerItem) current.getItem();
-                        FluidStack liquid = container.getFluid(current);
-                        FluidStack tankLiquid = tank.getTankInfo(ForgeDirection.UNKNOWN)[0].fluid;
-                        boolean mustDrain = liquid == null || liquid.amount == 0;
-                        boolean mustFill = tankLiquid == null || tankLiquid.amount == 0;
-                        if (mustDrain && mustFill) {
-                            // Both are empty, do nothing
-                        } else if (mustDrain || !entityplayer.isSneaking()) {
-                            liquid = tank.drain(ForgeDirection.UNKNOWN, 1000, false);
-                            int qtyToFill = container.fill(current, liquid, true);
-                            tank.drain(ForgeDirection.UNKNOWN, qtyToFill, true);
-                        } else if (mustFill || entityplayer.isSneaking()) {
-                            if (liquid.amount > 0) {
-                                int qty = tank.fill(ForgeDirection.UNKNOWN, liquid, false);
-                                tank.fill(ForgeDirection.UNKNOWN, container.drain(current, qty, true), true);
-                            }
+            // tank -> container logic
+            FluidStack available = clicked.getTankInfo(ForgeDirection.UNKNOWN)[0].fluid;
+            if (available != null) {
+                ItemStack preview = FluidContainerRegistry.fillFluidContainer(available, held);
+                FluidStack willContain = FluidContainerRegistry.getFluidForFilledItem(preview);
+                if (willContain != null) {
+                    if (!BuildCraftCore.debugWorldgen && !creative) {
+                        if (held.stackSize > 1) {
+                            if (!player.inventory.addItemStackToInventory(preview)) return false;
+                            player.inventory
+                                    .setInventorySlotContents(player.inventory.currentItem, InvUtils.consumeItem(held));
+                        } else {
+                            player.inventory
+                                    .setInventorySlotContents(player.inventory.currentItem, InvUtils.consumeItem(held));
+                            player.inventory.setInventorySlotContents(player.inventory.currentItem, preview);
```

Comment:
the first `setInventorySlotContents` is overridden by the second

also you changed `InvUtils.consumeItem(held)` to `held.splitStack(1)` in code above but not here